test/s3_test.go

```diff
@@ -0,0 +1,287 @@
+package test
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/gruntwork-io/terratest/modules/aws"
+	"github.com/gruntwork-io/terratest/modules/logger"
+	"github.com/gruntwork-io/terratest/modules/random"
+	"github.com/gruntwork-io/terratest/modules/retry"
+	"github.com/gruntwork-io/terratest/modules/terraform"
+	test_structure "github.com/gruntwork-io/terratest/modules/test-structure"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTerraformS3Bucket(t *testing.T) {
+	t.Parallel()
+
+	// For convenience - uncomment these as well as the "os" import
+	// when doing local testing if you need to skip any sections.
+	//
+	// A common usage for this would be skipping teardown first (keeping infrastructure)
+	// and then in your next run skip the setup* and create* steps. This way you can keep testing
+	// your Go test against your infrastructure quicker. Be mindful of random-ids, as they would be updated
+	// on each run, which would make some assertions fail.
+
+	// os.Setenv("SKIP_", "true")
+	// os.Setenv("TERRATEST_REGION", "us-east-1")
+
+	// os.Setenv("SKIP_setup_options", "true")
+	// os.Setenv("SKIP_create_bucket", "true")
+	// os.Setenv("SKIP_validate_bucket", "true")
+	// os.Setenv("SKIP_create_ro_objects", "true")
+	// os.Setenv("SKIP_setup_role_options", "true")
+	// os.Setenv("SKIP_create_role", "true")
+	// os.Setenv("SKIP_validate_bucket_and_policies", "true")
+
+	// os.Setenv("SKIP_teardown", "true")
+	// os.Setenv("SKIP_teardown_role", "true")
+
+	// string to be used as body for files to be created
+	testBodyString := "test"
+
+	// define struct for each object Test Case
+	type objectTestCase struct {
+		key             string
+		encryption      string
+		expectPassRead  bool
+		expectPassWrite bool
+	}
+
+	// list of different buckets that will be created to be tested
+	var bucketTestCases = []struct {
+		testName      string
+		pathRO        []string
+		pathRW        []string
+		objTestCases  []objectTestCase
+		sleepDuration int
+	}{
+		{
+			"TestBucket1",
+			[]string{"path/to/ro-folder"},
+			[]string{"path/to/rw-folder"},
+			[]objectTestCase{
+				{
+					key:             "path/to/ro-folder/obj1",
+					encryption:      "AES256",
+					expectPassRead:  true,
+					expectPassWrite: false,
+				},
+				{
+					key:             "path/to/rw-folder/obj2",
+					encryption:      "AES256",
+					expectPassRead:  true,
+					expectPassWrite: true,
+				},
+				{
+					key:             "other/folder/obj3",
+					encryption:      "AES256",
+					expectPassRead:  false,
+					expectPassWrite: false,
+				},
+			},
+			0,
+		},
+		{
+			"TestBucket2",
+			[]string{"other/path/to/ro-folder"},
+			[]string{"other/path/to/rw-folder"},
+			[]objectTestCase{
+				{
+					key: "path/to/ro-folder/obj1",
+					// encryption:      "",
```

Comment:
In Go we can't use `nil` for strings, as strings are internally treated as an immutable read-only value, not a pointer. 
There is a concept of "zero value" for different "types" in Go, `string`'s zero-vale is `""` (it is ´nil´ for struct objects, maps, slices etc.). That zero-value is the default of new variables. So, commenting this or not is interchangeable, as it will be an empty string (zero-value of the ´string´ type in Go).

Not setting encryption would cause all uploads to fail, hence the "expectWrite, expectRead = false".
This is the approach I used to make sure unencrypted objects cannot be uploaded to the bucket.

I am adding more comments on this, since it may cause confusion to code readers.